### PR TITLE
Show devices with template

### PR DESCRIPTION
**Pull request by linxt20**
- **Function Overview:** Upon using `show devices`, this feature enables viewing template information activated for each device, and allows querying a list of devices based on a specified template name using `where` clause.

- **Detailed Functionality:**
  - **Show devices Query for Template Usage:** Enhances the existing `show devices` display (devices, isAligned) by adding a new column (`template`) showcasing device usage of activated templates. For cases where a template is not used or hasn't been activated, it will uniformly display `null`. For devices with activated templates, the name of the activated template will be displayed.

  - **Filtering Devices Activated by a Specific Template using `where template=xxx`:** The presentation format is similar to that of `show devices`, based on the results of `show devices` (devices, isAligned, template). It filters the activation template information column based on the specified template name.

- **Expected Functionality:** // Statement Execution - Displaying the activated template column effect

```
IoTDB > show devices

// Execution Result
+-----------------+-------------+-------------+
|    devices      |  isAligned  |   template  |
+-----------------+-------------+-------------+
| root.ln.wf01.wt01 |    false    |     t1      |
| root.ll.wf02.wt02 |    false    |    null     |
+-----------------+-------------+-------------+
```

```
// Statement Execution - Filtering based on device column
IoTDB > show devices root.ln.**

// Execution Result
+-----------------+-------------+-------------+
|    devices      |  isAligned  |   template  |
+-----------------+-------------+-------------+
| root.ln.wf01.wt01 |    false    |     t1      |
+-----------------+-------------+-------------+
```

```
// Statement Execution - Filtering based on template column with equality condition
IoTDB > show devices where template = t1

// Execution Result
+-----------------+-------------+-------------+
|    devices      |  isAligned  |   template  |
+-----------------+-------------+-------------+
| root.ln.wf01.wt01 |    false    |     t1      |
+-----------------+-------------+-------------+
```

```
// Statement Execution - Filtering based on template column with inequality condition
IoTDB > show devices where template != t1

// Execution Result
+-----------------+-------------+-------------+
|    devices      |  isAligned  |   template  |
+-----------------+-------------+-------------+
| root.ll.wf02.wt02 |    false    |    null     |
+-----------------+-------------+-------------+
```

- **Implementing Feature Logic, Modifications Sequentially According to the MPP Query Process**
  - **G4 Grammar for Statement Generation**: Here, modifications are made within the IoTDBSqlParser in the devicesWhereClause, adding templateEqualExpression to support the syntax of filtering templates in the 'where' clause.

  - **Implementation of TemplateFilter Logic:** Corresponding to the G4 grammar, enhancements are made within the visitshowdevice of ASTVisitor to include parsing for templateEqualExpression content. This content is read as a SchemaFilter. Currently, there is no specific SchemaFilter implementation for templates. Hence, a TemplateFilter class is implemented in iotdb-core/node-commons under schema/filter. Additionally, the SchemaFilterFactory management class is updated with createTemplateIdFilter to uniformly create TemplateFilter instances. The SchemaFilterType class now includes TEMPLATE_FILTER type, and within the SchemaFilterVisitor class, the default visit for visitTemplateFilter is included.

  - **Parsing in QueryExecution:** In analyzeVisitor within visitshowdevice, when retrieving header information, the header for template columns is added.

  - **Execution by Operator:** In trygetnext, a schemaReader is first constructed. Within the getDeviceReader function of schemaReader, DeviceFilterVisitor is utilized for handling TemplateFilter access. Implementation for template filtering judgment concerning devices needs to be done within DeviceFilterVisitor. Once schemaReader construction is completed, setColumns is invoked to set the results in the corresponding columns. Modification is required in DeviceSchemaSource's transformToTsBlockColumns to append the device's template activation information at the end. Retrieval of template information from the device is necessary, thus, a new interface is added to IDeviceSchemaInfo, which is implemented within ShowDevicesResult.

### CODE REVIEW

File: iotdb-core/antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IoTDBSqlParser.g4 [hunk 1]
```diff
@@ -210,7 +210,11 @@ countNodes
 
 // ---- Timeseries Where Clause
 devicesWhereClause
-    : WHERE deviceContainsExpression
+    : WHERE (deviceContainsExpression | templateEqualExpression)
+    ;
+
+templateEqualExpression
+    : TEMPLATE (operator_eq | OPERATOR_NEQ)  templateName=identifier
```
Comment (MarcosZyk) [inline iotdb-core/antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IoTDBSqlParser.g4:217]: Why these two operator, "operator_eq" and "OPERATOR_NEQ", is different in character case? Make the code style consistent.

**Review by MarcosZyk — COMMENTED** [review]
PTAL~

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/common/header/ColumnHeaderConstant.java [hunk 2]
```diff
@@ -49,6 +49,8 @@ private ColumnHeaderConstant() {
   public static final String DEADBAND = "Deadband";
   public static final String DEADBAND_PARAMETERS = "DeadbandParameters";
   public static final String IS_ALIGNED = "IsAligned";
+  public static final String Template = "Template";
```
Comment (MiniSho) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/common/header/ColumnHeaderConstant.java:52]: Capitalize `Template` :)

Comment (Cpaulyz) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/common/header/ColumnHeaderConstant.java:52]: Keep the code style consistent with the original code.
```suggestion
  public static final String TEMPLATE = "Template";
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/schema/source/DeviceSchemaSource.java [hunk 3]
```diff
@@ -96,15 +97,28 @@ public void transformToTsBlockColumns(
     builder
         .getColumnBuilder(0)
         .writeBinary(new Binary(device.getFullPath(), TSFileConfig.STRING_CHARSET));
+
+    String TemplateName = "null";
```
Comment (Cpaulyz) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/schema/source/DeviceSchemaSource.java:101]: Do not use MAGIC NUMBER.

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/schema/filter/SchemaFilterFactory.java [hunk 4]
```diff
@@ -44,6 +45,10 @@ public static SchemaFilter createViewTypeFilter(ViewType viewType) {
     return new ViewTypeFilter(viewType);
   }
 
+  public static SchemaFilter createTemplateIdFilter(String templateName, boolean isEqual) {
```
Comment (Cpaulyz) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/schema/filter/SchemaFilterFactory.java:48]: Please let the name of the method express the exact meaning.
```suggestion
  public static SchemaFilter createTemplateNameFilter(String templateName, boolean isEqual) {
```

**Review by Cpaulyz — CHANGES_REQUESTED** [review]
How will the system react if the user executes such SQL? Please add some unit tests if this is not supported.

```
show devices where template=t1 and device contains 't';
show devices where device contains 't' and template=t1 ;
```

BTW, dont forget to update document in https://github.com/apache/iotdb-docs

File: integration-test/src/test/java/org/apache/iotdb/db/it/schema/IoTDBMetadataFetchIT.java [hunk 5]
```diff
@@ -292,6 +326,7 @@ public void showDevicesTest() throws SQLException {
               builder.append(resultSet.getString(i)).append(",");
             }
             String string = builder.toString();
+            System.out.println(string);
```
[the remark this answers is not in the document]
Comment (linxt20) [reply integration-test/src/test/java/org/apache/iotdb/db/it/schema/IoTDBMetadataFetchIT.java]: Fixed.

File: iotdb-core/antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IoTDBSqlParser.g4 (hunk 1 above)
Comment (linxt20) [reply iotdb-core/antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IoTDBSqlParser.g4]: Fixed.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/common/header/ColumnHeaderConstant.java (hunk 2 above)
Comment (linxt20) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/common/header/ColumnHeaderConstant.java]: Fixed.

Comment (linxt20) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/common/header/ColumnHeaderConstant.java]: Fixed.

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/schema/filter/SchemaFilterFactory.java (hunk 4 above)
Comment (linxt20) [reply iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/schema/filter/SchemaFilterFactory.java]: Fixed.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/utils/filter/DeviceFilterVisitor.java [hunk 6]
```diff
@@ -38,4 +40,26 @@ public boolean visitPathContainsFilter(
     }
     return info.getFullPath().toLowerCase().contains(pathContainsFilter.getContainString());
   }
+
+  @Override
+  public boolean visitTemplateFilter(TemplateFilter templateFilter, IDeviceSchemaInfo info) {
+    if (templateFilter.getTemplateName() == null) {
+      return true;
+    }
+    String TemplateName = "null";
```
Comment (MarcosZyk) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/utils/filter/DeviceFilterVisitor.java:49]: Use static variables.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/utils/filter/DeviceFilterVisitor.java [hunk 7]
```diff
@@ -38,4 +40,26 @@ public boolean visitPathContainsFilter(
...
+    boolean EqualAns;
```
Comment (MarcosZyk) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/utils/filter/DeviceFilterVisitor.java:50]: stay with camel style.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/utils/filter/DeviceFilterVisitor.java (hunk 6 above)
Comment (linxt20) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/utils/filter/DeviceFilterVisitor.java:49]: Fixed.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/utils/filter/DeviceFilterVisitor.java (hunk 7 above)
Comment (linxt20) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/utils/filter/DeviceFilterVisitor.java:50]: Fixed.

**Review by MarcosZyk — APPROVED** [review]

**Pull request merged**